plugins/out_s3/s3.c

```diff
@@ -1343,6 +1348,169 @@ static void cb_s3_upload(struct flb_config *config, void *data)
+    records = flb_mp_count(data, bytes);
+    if (records <= 0) {
+        return NULL;
+    }
+
+    /* Allocate buffer to store log_key contents */
+    val_buf = flb_malloc(msgpack_size);
+    if (val_buf == NULL) {
+        flb_plg_error(ctx->ins, "Could not allocate enough "
+                      "memory to read record");
+        flb_errno();
+        return NULL;
+    }
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&tmp_sbuf);
+
+    msgpack_unpacked_init(&result);
+    while (!alloc_error && 
+           msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        /* Each array must have two entries: time and record */
+        root = result.data;
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            continue;
+        }
+        if (root.via.array.size != 2) {
+            continue;
+        }
+
+        /* Get the record/map */
+        map = root.via.array.ptr[1];
+        if (map.type != MSGPACK_OBJECT_MAP) {
+            continue;
+        }
+        map_size = map.via.map.size;
+
+        /* Extract log_key from record and append to output buffer */
+        for (i = 0; i < map_size; i++) {
+            key = map.via.map.ptr[i].key;
+            val = map.via.map.ptr[i].val;
+
+            if (key.type == MSGPACK_OBJECT_BIN) {
+                key_str  = (char *) key.via.bin.ptr;
+                key_str_size = key.via.bin.size;
+                check = FLB_TRUE;
+            }
+            if (key.type == MSGPACK_OBJECT_STR) {
+                key_str  = (char *) key.via.str.ptr;
+                key_str_size = key.via.str.size;
+                check = FLB_TRUE;
+            }
+
+            if (check == FLB_TRUE) {
+                if (strncmp(ctx->log_key, key_str, key_str_size) == 0) {
+                    found = FLB_TRUE;
+
+                    /* 
+                     * Copy contents of value into buffer. Necessary to strncpy
+                     * strings because flb_msgpack_to_json does not handle nested
+                     * JSON gracefully and double escapes them.
+                     */
+                    if (val.type == MSGPACK_OBJECT_BIN) {
+                        strncpy(val_buf, val.via.bin.ptr, val.via.bin.size);
+                        val_buf[val.via.bin.size] = '\0';
+                    }
+                    else if (val.type == MSGPACK_OBJECT_STR) {
+                        strncpy(val_buf, val.via.str.ptr, val.via.str.size);
+                        val_buf[val.via.str.size] = '\0';
+                    }
+                    else {
+                        ret = flb_msgpack_to_json(val_buf, msgpack_size, &val);
+                        if (ret < 0) {
+                            break;
+                        }
+                    }
+
+                    /* Create or concatenate to out_buf for full record chunk */
+                    if (out_buf == NULL) {
+                        out_buf = flb_sds_create(val_buf);
```

Comment:
This code is very inefficient and I don't understand it. 


Why do you need both val_buf and out_buf?

With out_buf, you initialize it to the strlen of val_buf on the first iteration, and then you will realloc it frequently with the calls to `flb_sds_cat`. So you are not doing efficient memory allocations.

Remember that the goal was to avoid any allocation in the loop. 

I do not think you need two buffers. I think you can do this with only 1 buffer, allocated before the loop. You can track an offset for in that single buffer and repeatedly write the logs & newlines to it. No need to copy to another buffer.